r/tests/testthat/test-dataset.R

```diff
@@ -90,13 +90,39 @@ test_that("Hive partitioning", {
   expect_is(ds, "Dataset")
   expect_equivalent(
     ds %>%
+      filter(group == 2) %>%
       select(chr, dbl) %>%
       filter(dbl > 7 & dbl < 53) %>%
       arrange(dbl),
-    rbind(
-      df1[8:10, c("chr", "dbl")],
-      df2[1:2, c("chr", "dbl")]
-    )
+    df2[1:2, c("chr", "dbl")]
+  )
+})
+
+test_that("Partition scheme inference", {
+  # These are the same tests as above, just using the *PartitionSchemeDiscovery
+  ds1 <- open_dataset(dataset_dir, partition = "part")
+  expect_identical(names(ds1), c(names(df1), "part"))
+  expect_equivalent(
+    ds1 %>%
+      select(string = chr, integer = int, part) %>%
+      filter(integer > 6 & part == 1) %>%
+      summarize(mean = mean(integer)),
+    df1 %>%
+      select(string = chr, integer = int) %>%
+      filter(integer > 6) %>%
+      summarize(mean = mean(integer))
+  )
+
+  ds2 <- open_dataset(hive_dir)
+  expect_identical(names(ds2), c(names(df1), "group", "other"))
+  print(ds2$schema)
+  expect_equivalent(
```

Comment:
@bkietz this test fails, which suggests to me that the HivePartitionSchemeDiscovery isn't working fully. The fields do appear in the schema but the filtering isn't happening--the result has too many rows in it.

Reply:
fixed in https://github.com/apache/arrow/pull/6077/commits/6caaa8145fefd1ad510648559293ddc067e80b7b
HivePartitionSchemeDiscovery was constructing a SchemaPartitionScheme instead of a HivePartitionScheme. Wasn't caught because we only have unit tests for HivePartitionSchemeDiscovery in C++ (e2e uses SchemaPartitionSchemeDiscovery).